src/arithmetic/canonical_simplify.cc

```diff
@@ -0,0 +1,863 @@
+    this->args.emplace_back(std::move(other));
+  }
+
+  void AddToSelf(const SumExpr& other, int64_t scale);
+
+  static constexpr const char* _type_key = "arith.SumExpr";
+  TVM_DECLARE_NODE_TYPE_INFO(SumExprNode, CanonicalExprNode);
+
+ private:
+  /*!
+   * \brief Simplify the args by merging SplitExprs
+   * \param args The original list of arguments.
+   * \return simplified version.
+   */
+  static std::vector<SplitExpr>
+  SimplifySplitExprs(std::vector<SplitExpr> args) {
+    // NOTE: This algorithm relies on the factor that args are divided into segments
+    // and each segment is sorted in descending order of lower_factor.
+    for (size_t i = 0; i < args.size(); ++i) {
+      if (args[i]->scale == 0) continue;
+      for (size_t j = i + 1; j < args.size(); ++j) {
+        SplitExpr& lhs = args[i];
+        SplitExpr& rhs = args[j];
+        if (!lhs->IndexEqual(rhs)) break;
+        if (lhs->upper_factor < rhs->lower_factor) break;
+        if (lhs->lower_factor == rhs->upper_factor &&
+            lhs->scale % rhs->scale == 0 &&
+            lhs->lower_factor == (lhs->scale / rhs->scale) * rhs->lower_factor) {
+          // Rules used in the proof:
+          //
+          // Rule 1:  (x % (c * s)) / c  =  (x / c) % s
+          // Proof:
+          //  x can always be decomposed into p * c * s + q * c + r
+          //  where  0 <= q * c + r < c * s  and  0 <= r  <  c.
+          //  Then, lhs = ((p * c * s + q * c + r) % (c * s)) / c = (q * c + r) / c = q
+          //  rhs = ((p * c * s + q * c + r) / c) % s = (p * s + q) % s = q
+          //  Thus, lhs = rhs
+          //
+          // The above proof is for the floordiv.
+          // The same rule also holds for trucdiv(division rule in C).
+          // Because both sides only involve mul, div and mod,
+          // we can take abs of x, c and s, apply the floordiv proof,
+          // and finally add the sign back.
+          //
+          // Rule 2:  (x / s) * s + x % s = x  (true for both truc and floor div)
+          //
+          // General merge condition and proof:
+          // - x = lhs->index % lhs->upper_factor
+          // - s = lhs->scale / rhs->scale
+          // - c = rhs->lower_factor
+          //
+          //    ((x / (c * s)) * s + (x % (c * s)) / c
+          // => ((x / c) / s) * s + ((x / c) % s)
+          // => (x / c)
+          //
+          // Examples:
+          //
+          //    (z / 6) * 6 + ((z % 6) / 3) * 3
+          // => ((z / 6) * 2 + (z % 6) / 3) * 3
+          // => (z / 3) * 3
+          // note: x = z, c = 3, s = 2
+          //
+          //    ((z % 12) / 6) * 6 + ((z % 6) / 3) * 3
+          // => (((z % 12) / 6) * 2 + ((z % 12) % 6) / 3) * 3
+          // => ((z % 12) / 3) * 3
+          // note: x = z % 12, c = 3, s = 2
+          // note also the invariance lhs->upper_factor % lhs->lower_factor == 0
+          //
+          SplitExprNode* merged = rhs.CopyOnWrite();
+          merged->upper_factor = lhs->upper_factor;
```

Comment:
Is this correct only when lhs->uppper_factor == kPosInf?  For example, `((x % 5) / (3 * 2)) * 2 + (x % (3 * 2)) / 3` is simplified to `x % 5 / 3`, but this is not correct when x == 5.

Reply:
See comment above in invariance

Reply:
Ah I see, sorry for my confusion.